tests/utils.py

```diff
@@ -74,3 +77,117 @@ def wait_for_server(host, port, timeout=10):
+
+      port:
+        The port used to access the server. If none is provided,
+        then one will be generated.
+        Default is None.
+
+      timeout:
+        Time in seconds in which the server should start or otherwise
+        TimeoutError error will be raised.
+        If 0 is given, no check if the server has started will be done.
+        Default is 10.
+
+      popen_cwd:
+        Current working directory used when instancing a
+        subprocess.Popen object.
+        Default is "."
+
+      extra_cmd_args:
+        List of additional arguments for the command
+        which will be ran from the subprocess.
+        More precisely "python -u <path_to_server> <port> <extra_cmd_args>".
+        Default is empty list.
+  """
+
+  # Starts subprocess which runs the server,
+  # sets up the TempFile, waits for server response.
+  def __init__(self, log, server='simple_server.py',
+      port=None, timeout=10, popen_cwd=".",
+      extra_cmd_args=[]):
+    # Create temporary log file used for logging stdout and stderr
+    # of the subprocess. In the mode "r+"" stands for reading and writing
+    # and "t" stands for text mode.
+    self.__temp_log_file = tempfile.TemporaryFile(mode='r+t')
+
+    self.server = server
+    self.port = port or random.randint(30000, 45000)
+    self.__logger = log
+
+    # The "-u" option forces stdin, stdout and stderr to be unbuffered.
+    command = ['python', '-u', server, str(self.port)] + extra_cmd_args
+
+    # We are reusing one server subprocess in multiple unit tests, but we are
+    # collecting the logs per test.
+    self.__server_process = subprocess.Popen(command,
+        stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
+
+    self.__logger.info('Server process started.')
+    self.__logger.info('Server process id: ' + str(self.__server_process.pid))
+    self.__logger.info('Serving on port: ' + str(self.port))
+
+    if timeout > 0:
+      try:
+        wait_for_server('localhost', self.port, timeout)
+      except Exception as e:
+        self.__logger.error(
+            "Error while waiting for the server to start: " + repr(e))
+        # Make sure that errors from the server side will be logged.
+        self.flush_log()
+        raise e
+
+
+
+  # Logs contents from TempFile, truncates buffer
+  def flush_log(self):
+
+    # Seek is needed to move the pointer to the beginning of the file, because
+    # the subprocess could have read and/or write and thus moved the pointer.
+    self.__temp_log_file.seek(0)
+    log_message = self.__temp_log_file.read()
+
+    if len(log_message) > 0:
+      self.__logger.info("*******************")
+
+      self.__logger.info("Logs from the subprocess running server: " \
+          + self.server)
+
+      messages = log_message.split('\n')
+      for line in messages:
+        if line != "":
+          self.__logger.info("INFO: " + line)
+
+      self.__logger.info("*******************")
```

Comment:
I still have issues with this bit:
* very verbose: most of the time the log itself is a single line: this prints 4 times as much
* should not print "INFO: ": this is what the log level is for. If the user wants it printed, they will use a log formatter that prints the log level
* I think the earlier version of a single log entry for the server log was better (it can still be multiline) as that can be made to clearly show the beginning and end of server log without need of extra lines of "************"  (I'll paste the example I tested with below) -- do you see a problem with this approach?

```
    # Build a single log entry with multiline content
    title = "Test server (" + self.server + ") output:"
    message = [title] + log_message.splitlines()
    logger.info('\n| '.join(message))
```

This way the normal oneliner now looks like this (assuming basicConfig default formatter)

```
INFO:logger_name:Test server (simple_server.py) output:
| code 404, message File not found
```

Reply:
I fixed it in an upcoming commit.